### PR TITLE
BE: Request Model

### LINKED ISSUE #9: BE: Request Model
The following are properties you may want to consider adding to this model but is not limited to:

\- user\_id (user who is requesting)
\- sitter\_id (pet sitter user id)
\- start (date requested to start)
\- end (date requested to end)
\- accepted (default False) 
\- declined (default False)
\- paid (default False)

### PR DESCRIPTION
**Pull request by ChadNaps**
### What this PR does (required):
- Add request model to project

### Any information needed to test this feature (required):
Note: Info also available in README.md
- `cd server`
- `npm test`
- `npm run coverage`

### Any issues with the current functionality (optional):
- I tried to consider all the possible things we might need in the request model, it's possible I missed something however.
- In addition to the specs, I added an address submodel that stores
    - Pickup Address: The address the dog will be picked up from
    - Service Address: The address the dog will be cared for at
    - Dropoff Address: The address the dog will be dropped off at

### CODE REVIEW

File: server/models/Request.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  },
+  paid: {
+    type: Boolean,
+    required: true,
+    default: false,
+  },
+  address: addressSchema,
+});
+
+module.exports = Request = mongoose.model("Request", requestSchema);
```
Comment (sundayezeilo) [inline server/models/Request.js]: It'd be good to separate `Request` model and `Address` model in different files.

**Review by sundayezeilo — COMMENTED** [review]
Good work with this one! One comment to have a look, though!

File: server/models/Request.js (hunk 1 above)
Comment (ChadNaps) [reply server/models/Request.js]: Good note, I'll get that done. Thanks!

**Review by sundayezeilo — COMMENTED** [review]
Thank you for making the requested changes :+1:

File: server/models/Address.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  },
+  dropoff: {
+    type: String,
+    default: function () {
+      if (this.pickup) {
+        return this.pickup;
+      }
+
+      return null;
+    },
```
Comment (moffatethan) [inline server/models/Address.js]: Wouldn't this just be unnecessary? Since we're just using the pickup address?

File: server/models/Address.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  pickup: {
+    type: String,
+    required: true,
+    default: null,
+  },
+  service: {
+    type: String,
+    required: true,
+    default: null,
+  },
```
Comment (moffatethan) [inline server/models/Address.js]: Not sure what this is for?

File: server/models/Request.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    type: Boolean,
+    required: true,
+    default: false,
+  },
+  paid: {
+    type: Boolean,
+    required: true,
+    default: false,
+  },
+  address: addressSchema,
```
Comment (moffatethan) [inline server/models/Request.js:38]: I like the idea of having an address but the schema is not what you would expect the address to be. No street numbers or anything? 

Could we make that schema an actual address schema. I'm not worried about pickup or drop-off particularly.

**Review by moffatethan — CHANGES_REQUESTED** [review]
Overall good job with this, just some concerns mainly on the address schema.

File: server/models/Request.js (hunk 4 above)
Comment (moffatethan) [reply server/models/Request.js:38]: Also, I think we can just have the customer enter an address and then we can just input that in there automatically.

File: server/models/Address.js (hunk 2 above)
Comment (ChadNaps) [reply server/models/Address.js]: Well, it's more-so to allow people to specify a different dropoff location, but I figured most would likely just have the dropoff be the same as the pickup. So, most of the time it wouldn't be necessary. But, if someone wanted to pick a different dropoff, the option is there.

That being said, I'm not married to it. If you want me to scrap it, can do!

File: server/models/Address.js (hunk 3 above)
Comment (ChadNaps) [reply server/models/Address.js]: It's the address the service will be rendered at. For example, taking care of puppers at the client's home vs the sitter's home vs maybe some kind of dog daycare facility.

File: server/models/Request.js (hunk 4 above)
Comment (ChadNaps) [reply server/models/Request.js:38]: That's... I'm not sure why I didn't think of it that way, lol. I envisioned the address just being a string like, "123 Test St, Calgary, AB, H0H 0H0". But yeah, that's a much better way of doing it. I'll get it changed around!

Comment (moffatethan) [reply server/models/Request.js:38]: Yeah, I think ideally we just have them enter it in the settings then we just use that (maybe in the future can have multiple addresses)

File: server/models/Address.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  postalCode: {
+    type: String,
+    maxlength: 7,
+    required: true,
+    validate: {
+      validator: function (value) {
+        // Allows: h2t-1b8; h2z 1b8; H2Z1B8
+        // Disallows: Z2T 1B8 (leading Z); H2T 1O3 (contains O); Leading Z,W or to contain D, F, I, O, Q or U
+        // Credit: https://stackoverflow.com/a/46761018
```
Comment (moffatethan) [inline server/models/Address.js]: Other two comments I'm fine with, credit is great but this can be removed 😅. Thanks Chad!

**Review by moffatethan — APPROVED** [review]
Approved with one little clean up comment!

**Pull request merged**